dockertest/subtestbase_unittests.py

```diff
@@ -98,41 +85,87 @@ def test_failif_redhat(self):
+class TestFailIfNotIn(SubtestBaseTestBase):
     """
     Tests for failif_not_in()
     """
-    pass
-
 
-# Generate tests for each case:
-#  https://stackoverflow.com/questions/32899/how-to-generate-dynamic-parametrized-unit-tests-in-python
-def test_generator_pass(needle, haystack):
-    def test(self):
-        import subtestbase
-        subtestbase.SubBase.failif_not_in(needle, haystack)
-    return test
+    # In each of these, the left-hand string(s) will be present at right.
+    expect_pass = [
+        ['a',        'a'],
+        ['a',        'aa'],
+        ['a',        'abc'],
+        ['a',        'cba'],
+        ['a|a',      'a'],
+        ['a | a',    'a'],
+        ['string',   'ceci nest pas une string'],
+        ['no|yes',   'googlyeyes'],
+        ['no | yes', 'googlyeyes'],
+        ['needle',   'needle in a haystack'],
+        ['needle',   'haystack with a needle in the middle'],
+        ['needle',   'haystack with, at end, a needle'],
+    ]
+
+    # The left-hand string(s) will NOT be in the right
+    expect_fail = [
+        ['a',         'b'],
+        ['needle',    'haystack'],
+        ['a|b|c',     'd'],
+        ['a | b | c', 'd'],
+        ['a | a | a', 'b'],
+    ]
+
+    def test_pass(self):
+        for needle, haystack in self.expect_pass:
+            # Automaticly shows needle & haystack in any failure message
+            with self.subTest(needle=needle, haystack=haystack):
+                result = self.stbsb.failif_not_in(needle, haystack)
+                self.assertIsNone(result)
+
+    def test_fail(self):
+        for needle, haystack in self.expect_fail:
+            with self.subTest(needle=needle, haystack=haystack):
+                self.assertRaises(DockerTestFail,
+                                  self.stbsb.failif_not_in,
+                                  needle, haystack)
+
+
+class TestKnownFailures(SubtestBaseTestBase):
+    """Tests known_failures() and SubBase.is_known_failure()"""
 
+    def setUp(self):
+        super(TestKnownFailures, self).setUp()
+        self.tmpfile = None
+        self.fake_open = lambda *args, **dargs: self.tmpfile
+        self.subtestbase.open = self.fake_open
+
+        self.docker_rpm = 'docker'
+        self.subtestbase.docker_rpm = lambda: self.docker_rpm
+
+    def tearDown(self):
+        # Don't let mocks or state leak anywhere
+        del self.subtestbase.open
+        del self.subtestbase.docker_rpm  
+        del self.subtestbase
+        super(SubtestBaseTestBase, self).tearDown()
+
+    def reset_tmpfile(self, contents=None):
+        self.tmpfile = TemporaryFile()
+        if contents:
+            self.tmpfile.write(contents)
+            self.tmpfile.seek(0,0)
+
+    def test_empty_known_failures(self):
+        self.reset_tmpfile()
+        known = self.subtestbase.known_failures()
+        self.assertEqual(known, {})
+
+    def test_known_failures(self):
+        self.reset_tmpfile('"one","two","three"\n')
```

Comment:
ya, let's do that.   it could use some more tests as well, but that was the best I could do at the time.